### ISSUE TITLE
Request Device: Ceiling Fan

### ISSUE THREAD

**Issue by givemelove** [athombv/com.tuya#8]
Looking to get Ceiling Fans added as a supported device. I have 4 devices (same make / model). They are a combo ceiling fan and lamp. Here's an [Amazon link for the item](https://www.amazon.com/gp/product/B0CJDXS1JD)

The core functionality would be to be able to operate the fan (speeds) and the light (color and dimmer).

There are some features that could be a plus but may be overkill for a v1.

- change fan direction (summer / winter)
- My remote has a "wind" effect mode where the module would change the intensity following a pattern.

Here's the info for one device, with the key and PII info removed:

```json
{"active_time":1717640325,"biz_type":0,"category":"fs","create_time":1715646363,"icon":"smart/icon/ay1540882177261iTAUO/55bba57fc88d7744bcdc8f541384b126.jpg","id":"string","ip":"ip_address","lat":"numeric","local_key":"string","lon":"numeric","model":"SF-52","name":"One","online":true,"owner_id":"196613604","product_id":"t0qzrmur7pyzjuco","product_name":"","status":[{"code":"switch","value":false},{"code":"fan_speed_percent","value":6},{"code":"light","value":true},{"code":"bright_value","value":100},{"code":"temp_value","value":0}],"sub":false,"time_zone":"-05:00","uid":"string","update_time":1717676232,"uuid":"string"}
```

**Comment by gslender** [comment]
Have added a driver for Fan and got onOff Capability working in a fork for anyone who is interested. 

https://github.com/gslender/com.tuya

I will probably add speed control, but recognise there is no inbuilt capability for Homey for fan speed. My Tuya fans support 6 speeds but report them (oddly) as a % between 0 and 100, but only respond to 0,1,2,3,4,5,6 as the functional values that work !?

**Comment by givemelove** [comment]
@gslender I can confirm I can see my 4 fans as well. The fan on/off seems to work great

**Comment by gslender** [comment]
Have updated the app on my fork to include dim/level capability restricted to 1-6 with 1 value steps. Note sure how Athom will attempt to address this as there is no real way to know how many values the fan will support. Doesn't look like Tuya provide that information or maybe it is a bug in their API. It works, but if your fan only support 4 speeds (or has more than 6) then what I've coded won't work for you

**Comment by WeeJeWel** [comment]
Fixed by https://github.com/athombv/com.tuya/pull/14.

**Comment by givemelove** [comment]
Hey folks,

While the fan portion works, I was wondering if we have the ability to address the light piece. 

In the original request, the status key is an array that contains multiple capabilities (Switch, Brightness, Fan Switch, Fan Speed). While the latter two were added, is there a way to add the light switch component as well?

**Comment by gslender** [comment]
@givemelove if you send the status console output, I would be willing to look at adding it to my fork, you test and confirm it works, and then I can submit another PR if @WeeJeWel is happy with that approach

**Comment by WeeJeWel** [comment]
Sounds good!

**Comment by givemelove** [comment]
Thanks, I appreciate it. Happy to help in any way.

@gslender what do you need exactly? You need to to run the latest integration locally and give you some logs once I trigger a device, is that it?

**Comment by gslender** [comment]
@givemelove is the detail in the OP as per the missing features of the fan/light?

**Comment by gslender** [comment]
@givemelove I would also need you to switch to my fork at https://github.com/gslender/com.tuya/ to test and confirm that the changes are correct - so whilst I have a Tuya fan, mine don't have the light features

**Comment by gslender** [comment]
Actually, @WeeJeWel  would you mind thinking about how you wish a combo fan-light would be respresented in Homey? I have had this same issue with a wall switch that was a combo fan and light switch (zigbee) and I personally prefer having a single device that is able to (via an enum have fan off, low, medium, high) leaving the on/off for the light. In this specific Tuya case it also has a dimmer so that's adding another dimmer capabilty assuming the fan is pushed into an enum. I think it would be worthwhile thinking about a specific fan capability and maybe even a specific UI control that allows for on/off and levels (number or text labelled) that can be a slider etc.

@givemelove what's your thoughts ??

**Comment by gslender** [comment]
@WeeJeWel can you reopen this please too?

**Comment by givemelove** [comment]
@gslender Yeah, I can switch to your fork, that's not a problem. Yes, the missing features should be in my OP. The device I listed there has both capabilities.

With regards to your latter post re: how it would be represented in Homey, personally I may have a slight preference towards the multiple devices. While a single device could be "easier" for understanding that it's a single unit in the GUI, having a device per feature could also be beneficial for someone who uses the gui to trigger devices. I personally plan on using the fan with some flow rules so it doesn't really matter to me.

with that being said, I'm not familiar with other Tuya devices that may have multiple capabilities. If there's a device with 4 or so capabilities, it may become cumbersome to render all of that on the same card.

**Comment by gslender** [comment]
Sorry @givemelove I'm waiting on @WeeJeWel to provide some guidance on how to attack a device that has combo light and fan. Also, having a fan specific capability and UI would also help to ensure we can turn the fan on/off/low/med/high along with custom levels if the fan speed can be 1 to 6 or similar. I think adding the fan UI so that you can have a combo onOff, dim, fan capability would be ideal

**Comment by WeeJeWel** [comment]
Hey! Yeah this is a challenge 😅 The way I see it there are two options forward:

1. Add multiple capabilities to the same device, e.g. `onoff.light`.
2. Create a separate driver with the `light` class and let the user add two devices separately.

I would prefer option 2, because it works better with the 'Lights' control in the mobile app, and also speech commands such as 'Turn off all lights'.

We do need to change a bit in the Webhooks mechanism to allow two devices with the same ID receive those events though.

**Comment by gslender** [comment]
What about adding a specific Capability and UI for fan ??

**Comment by WeeJeWel** [comment]
That's something on the roadmap, but not really relevant if we want to have this working in the next few days :)

**Comment by gslender** [comment]
Fair point. Leave with me to have a play

@givemelove I will reach out again when I've got a commit in my fork that you can try and provide feedback 

@WeeJeWel any chance to reopen the issue?

**Comment by gslender** [comment]
@givemelove jump over to my fork at https://github.com/gslender/com.tuya/ and run the app as `homey app run -r` and report any issues or problems along with the output in the console.

You will need to delete and create a new device (a Fan Light Combo) and this should work if my guess about what's required is correct. Note that I've used the volume_mute and volume_set controls to hopefully mimic something that will allow you to control the fan, leaving the onoff and dim for the light.

**Comment by givemelove** [comment]
@gslender Apologies for the delayed response. The capabilities you implemented work. I can control both lighting and the fan on my devices. Thank you

**Comment by gslender** [comment]
@WeeJeWel the new PR I've raised would be sufficient (initially) for the multi-fan-light device to operate, though I'd recommend you consider adding a specific UI and fan-capabilities in the long term

**Comment by gslender** [comment]
@WeeJeWel have noticed your team replaced the fan icon (which was a ceiling fan pic) to a desk fan, and now with the limitted range of icons you provide in Homey, I can't set the icon back to something that resembles the original device.

Can you please, please, please think beyond fans being just desk fans and perhaps start supporting ceiling fans - the world is full of them and I don't own a single desk fan, but have 10 different types of ceiling fans in my home (as do most folks I know in my part of the world).

**Comment by bobvandevijver** [comment]
When 1.2.0 is released we will be looking into this a bit more (the new release actually contains a small fix for the fan driver (#123)). Are there any major problems left, except for the fan/light combination?

Regarding the icon: as we add drivers based on their functionality the icons we use will not always match with the device you have. We are looking into using different icons per device type, but we are still limited to the device types tuya uses obviously.

**Comment by bobvandevijver** [comment]
I'm not saying that it is a reasonable icon for a ceiling fan, I am saying that from the device category it cannot be concluded whether it is a standing or ceiling fan. Both are using the same `fd` (Fan) category. There is also an `fsd` (Ceiling fan light) category which seems to be a better fit for a ceiling fan with light support, but concluding from the specification posted by @givemelove that isn't used for the device opened here. 

Using the standing fan as icon is therefor nothing more than a design choice, but we do need to update the device image on the store to match the icon. That will be done in a future release.

For now, I would ask the both of you (@gslender and @givemelove) to post an update device specification which can be retrieved with the latest Tuya app version from the settings when the device is added as an other device. The question is whether we can correctly detect the light capability from that information or not. Please also note whether you actually have a light in your fan or not.

**Comment by gslender** [comment]
The previous posts provide all the details needed - as explained several times, the Tuya API does not make any distinction between a device that does or doesn't have a light as a fan device. The feature is described as an API but physically missing in devices that don't have the light. 

As such, the only solution, as I've painfully tried to explain but nobody is listening, is to create a separate and independent FanLight device that the user must choose.

As the for icon, as you include a desk/standing icon on your standard option choice for any device, it stands to commonsense and logic to include a ceiling fan icon in the Tuya App so that users with a ceiling fan can use that, and those with a desk fan can switch to the included icon in the Homey app. 

Is that reasonable?

**Comment by gslender** [comment]
> Here is my info for one of my 4 fans...
> 
> ```
> 
> {
> 
>   "active_time": 1717647195,
> 
>   "biz_type": 0,
> 
>   "category": "fs",
> 
>   "create_time": 1717647195,
> 
>   "icon": "smart/icon/ay1535601660405KA72f/50bfec90efcce440ec075e664d894f76.png",
> 
>   "id": "bf6fb59e48cf4a6ccckunc",
> 
>   "ip": "########",
> 
>   "lat": "-26.41",
> 
>   "local_key": "###########",
> 
>   "lon": "153.11",
> 
>   "model": "MZ-DCFAN",
> 
>   "name": "Lounge Fan",
> 
>   "online": true,
> 
>   "owner_id": "34319071",
> 
>   "product_id": "agyqzsfpafsvkrl9",
> 
>   "product_name": "DC吊扇",
> 
>   "status": [
> 
>     {
> 
>       "code": "switch",
> 
>       "value": false
> 
>     },
> 
>     {
> 
>       "code": "fan_speed_percent",
> 
>       "value": 2
> 
>     },
> 
>     {
> 
>       "code": "light",
> 
>       "value": false
> 
>     },
> 
>     {
> 
>       "code": "bright_value",
> 
>       "value": 1000
> 
>     },
> 
>     {
> 
>       "code": "temp_value",
> 
>       "value": 550
> 
>     }
> 
>   ],
> 
>   "sub": false,
> 
>   "time_zone": "+10:00",
> 
>   "uid": "eu1626678725509Ny23z",
> 
>   "update_time": 1717647235,
> 
>   "uuid": "b0c10c799234a73e"
> 
> }
> 
> ```

@bobvandevijver so you read this post?

**Comment by bobvandevijver** [comment]
@gslender Yes, which is why I asked to retrieve the latest one from the app.

**Comment by gslender** [comment]
Also, I've read the Tuya API and also the same issue exists in HomeAssistant support for Tuya fans. You can't detect, via the API, if a light is used/exists or not.

**Comment by gslender** [comment]
> So either your device has just been wrongly configured by Tuya/its brand (you never posted which one it actually is), or it actually has a light somewhere you are not aware of.

I'm not the person with the Light. I just forked this repo and made it work for @givemelove 

All that's happened since is that further updates to my PR have made my fan have the wrong icon (thanks a lot for doing that - sarcasm) and you've failed to implement any solution for the users with a light and fan combo.

Honestly, it's this kind of disregard for people in the community that makes people not bother contributing... it's been a total waste of my time ☹️

Best of luck and hope you figure it out - I'll leave it to you

**Comment by gslender** [comment]
@givemelove my fork has been sync'd again with the current Athom codebase, so if you wish to provide the output @bobvandevijver needs but retain the working fan-light device/drivers you can do so

**Comment by gslender** [comment]
@bobvandevijver below is the updated spec output. 

I have eight (8) ceiling fans from 3 different Tuya white-labeled vendors. 

They all report as being a 'fs' type and all report as having a light in the API but don't actually have a light in the device itself. 

My understanding is that Tuya don't make it mandatory that fan manufactures do or don't include the light hardware. They economise by making the software generic and turning the light on/off has no effect. I think you will eventually find that Tuya do this a lot in many of their devices - the API provides broad support but the device is sold with physical/electronic aspects disabled or missing depending on the model purchased etc.

Obviously if you make the Homey device report as having a light (because the Tuya API does), that's not only silly , but also annoying considering you don't have to do that - you could, as I've done, make an optional version of this device that is called FanLight and the user simply picks the most valid device that matches the hardware features they have. This is what I've been proposing and so far you've all just ignored, and why I'll just continue to use my Fork with the correct icon and light feature separated. I hope that you and Athom can internally discuss this fully and make an effort to think this through. 

Not everyone in the world has desk fans (I don't own a single one) and many homes in warm climates have multiple ceiling fans, and some folks even have lights in them... but not that many that you'd just assume all have lights.

```
{
    "device": {
        "active_time": 1717646605,
        "biz_type": 0,
        "category": "fs",
        "create_time": 1717646605,
        "icon": "smart/icon/ay1535601660405KA72f/50bfec90efcce440ec075e664d894f76.png",
        "id": "bff4cd2ed968b03942itau",
        "ip": "<redacted>",
        "lat": "<redacted>",
        "local_key": "<redacted>",
        "lon": "<redacted>",
        "model": "MZ-DCFAN",
        "name": "Dinning Table Fan",
        "online": true,
        "owner_id": "<redacted>",
        "product_id": "agyqzsfpafsvkrl9",
        "product_name": "DC吊扇",
        "status": [
            {
                "code": "switch",
                "value": false
            },
            {
                "code": "fan_speed_percent",
                "value": 2
            },
            {
                "code": "light",
                "value": false
            },
            {
                "code": "bright_value",
                "value": 1000
            },
            {
                "code": "temp_value",
                "value": 550
            }
        ],
        "sub": false,
        "time_zone": "+10:00",
        "uid": "<redacted>",
        "update_time": 1723222881,
        "uuid": "<redacted>"
    },
    "specifications": {
        "category": "fs",
        "functions": [
            {
                "code": "switch",
                "desc": "{}",
                "name": "风扇开关",
                "type": "Boolean",
                "values": "{}"
            },
            {
                "code": "fan_speed_percent",
                "desc": "{\"min\":\"1\",\"max\":\"6\",\"scale\":\"0\",\"step\":\"1\"}",
                "name": "无极风速",
                "type": "Integer",
                "values": "{\"min\":\"1\",\"max\":\"6\",\"scale\":\"0\",\"step\":\"1\"}"
            },
            {
                "code": "light",
                "desc": "{}",
                "name": "灯光开关",
                "type": "Boolean",
                "values": "{}"
            },
            {
                "code": "bright_value",
                "desc": "{\"min\":\"10\",\"max\":\"1000\",\"scale\":\"0\",\"step\":\"1\"}",
                "name": "亮度值",
                "type": "Integer",
                "values": "{\"min\":\"10\",\"max\":\"1000\",\"scale\":\"0\",\"step\":\"1\"}"
            },
            {
                "code": "temp_value",
                "desc": "{\"min\":\"0\",\"max\":\"1000\",\"scale\":\"0\",\"step\":\"1\"}",
                "name": "冷暖值",
                "type": "Integer",
                "values": "{\"min\":\"0\",\"max\":\"1000\",\"scale\":\"0\",\"step\":\"1\"}"
            }
        ],
        "status": [
            {
                "code": "switch",
                "name": "风扇开关",
                "type": "Boolean",
                "values": "{}"
            },
            {
                "code": "fan_speed_percent",
                "name": "无极风速",
                "type": "Integer",
                "values": "{\"min\":\"1\",\"max\":\"6\",\"scale\":\"0\",\"step\":\"1\"}"
            },
            {
                "code": "light",
                "name": "灯光开关",
                "type": "Boolean",
                "values": "{}"
            },
            {
                "code": "bright_value",
                "name": "亮度值",
                "type": "Integer",
                "values": "{\"min\":\"10\",\"max\":\"1000\",\"scale\":\"0\",\"step\":\"1\"}"
            },
            {
                "code": "temp_value",
                "name": "冷暖值",
                "type": "Integer",
                "values": "{\"min\":\"0\",\"max\":\"1000\",\"scale\":\"0\",\"step\":\"1\"}"
            }
        ]
    }
}
```

**Comment by bobvandevijver** [comment]
@gslender Thank you for the information! I was really hoping that the function specification would indicate that it doesn't allow for it to be controlled, but sadly it doesn't.

> My understanding is that Tuya don't make it mandatory that fan manufactures do or don't include the light hardware.

Tuya offers a configuration based hard/software platform, where it is up the manufacturer to request/configure (together with Tuya) which features are available for each product. It seems that they have been lazy with this specific product, which is a shame. We won't be adding a dedicated driver as that would return the same device with multiple drivers (which would introduce other issues), but we will make sure to add a setting with which the light capabilities can be easily removed from the device once added. We have something similar for the power measurements in the sockets, where not every manufacturer seems to adhere the documented power scaling.

Regarding the icon: I have already put in the request to make a ceiling fan icon available in the app, so you should be able to choose that easily in the future. By the way, did you see that Tuya also uses the default table fan icon for your device? https://images.tuyaeu.com/smart/icon/ay1535601660405KA72f/50bfec90efcce440ec075e664d894f76.png 🥲

**Comment by givemelove** [comment]
Sorry folks, 1.2.0 crashed my entire instance. Backups aren't helping. I'm literally rebuilding from scratch. It will take me a few days until I can retest anything.

**Comment by bobvandevijver** [comment]
I am closing this as with 1.3.0 we have rewritten fan support. If there are any other issues with that version please make sure to open a new ticket.
